### PR TITLE
initial SETUP-command

### PR DESCRIPTION
**Pull request by auamod**

### CODE REVIEW

File: vpn-daemon/main.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ func handleClientConnection(clientConnection net.Conn) {
 	managementIntPortList := []int{}
 	setPortsRegExp := regexp.MustCompile(`^SET_PORTS [0-9]+( [0-9]+)*$`)
 	disconnectRegExp := regexp.MustCompile(`^DISCONNECT [a-zA-Z0-9-.]+( [a-zA-Z0-9-.]+)*$`)
-	setupRegExp := regexp.MustCompile(`^SETUP ([a-zA-Z0-9-.]+) \[([a-zA-Z0-9-.]*(?:,[a-zA-Z0-9-.]+)*)\] (.{15,35})$`)
+	setupRegExp := regexp.MustCompile(`^SETUP ([a-zA-Z0-9-.]+) \[([a-zA-Z0-9-.]*(?:,[a-zA-Z0-9-.]+)*)\]$`)
```
Comment (fkooman) [inline vpn-daemon/main.go]: `SETUP_OPENVPN`

Comment (fkooman) [reply vpn-daemon/main.go]: The regexp does not seem to be correct... did you test this? I don't see any tests added?

`\[([a-zA-Z0-9-.]*(?:,[a-zA-Z0-9-.]+)*)\]`

The first `*` seems incorrect, has to be `+`? Otherwise this would match as well: `[,abc]`?

File: vpn-daemon/main.go [hunk 2]
```diff
@@ -153,17 +153,11 @@ func handleClientConnection(clientConnection net.Conn) {
...
-				writer.Flush()
-				fmt.Println(err.Error())
-				continue
-			}
+			commonName, profileList := prepareSetupParameters(setupRegExp.FindStringSubmatch(text)[1:])
 
 			cnProfileDir := filepath.Join("./c", commonName)
-			if err = os.RemoveAll(cnProfileDir); err != nil {
-				writer.WriteString(fmt.Sprintf("ERR: UNABLE_TO_REMOVE_EXISTING_CN_FILES\n"))
+			if err := os.RemoveAll(cnProfileDir); err != nil {
```
Comment (fkooman) [inline vpn-daemon/main.go]: The idea is to remove all CNs and then add them back (one by one) again? That's quite nice I guess...

File: vpn-daemon/main.go [hunk 3]
```diff
@@ -175,17 +169,15 @@ func handleClientConnection(clientConnection net.Conn) {
 				continue
 			}
 
-			err = os.MkdirAll(cnProfileDir, 0777)
-			if err != nil {
-				writer.WriteString(fmt.Sprintf("ERR: UNABLE_TO_CREATE_CN_DIRECTORY\n"))
+			if err := os.MkdirAll(cnProfileDir, 0755); err != nil {
```
Comment (fkooman) [inline vpn-daemon/main.go]: Actually nobody needs to be able to list directories, so I guess 0711 would work as well...

File: vpn-daemon/main.go [hunk 4]
```diff
@@ -302,20 +287,14 @@ func parseManagementPortList(managementStringPortList []string) ([]int, error) {
 	return managementIntPortList, nil
 }
 
-func prepareSetupParameters(setupStringParameters []string) (string, []string, time.Time, error) {
+func prepareSetupParameters(setupStringParameters []string) (string, []string) {
```
Comment (fkooman) [inline vpn-daemon/main.go]: I guess it is more "parse" than "prepare"...

File: vpn-daemon/main.go (hunk 1 above)
Comment (auamod) [reply vpn-daemon/main.go]: You're correct, this will accept `[,abc]`. However if its changed to `+`, `[]` will never be accepted

File: vpn-daemon/main.go (hunk 2 above)
Comment (auamod) [reply vpn-daemon/main.go]: It will only remove 1 CN (with its respective profiles)
This was handy when expiry_time was still around, instead of going through all the files and checking the file contents we could just delete everything and add them anew.

File: vpn-daemon/main.go (hunk 1 above)
Comment (auamod) [reply vpn-daemon/main.go]: `\[([a-zA-Z0-9-.]+(?:,[a-zA-Z0-9-.]+)*)*\]` does fit all description if I'm not mistaken

Comment (fkooman) [reply vpn-daemon/main.go]: okay... it seems we are reaching the point where regexps offer a diminishing return... can we think of a simpler command format so we don't need to have this crazy regexp but can instead use something simple? 

    SETUP_OPENPVN ${CN} ${P1} ${P2} ${P3}

What do you think?

Comment (auamod) [reply vpn-daemon/main.go]: In that case it will look like this `^SETUP_OPENVPN [a-zA-Z0-9-.]+(?: [a-zA-Z0-9-.]+)*$`
Functionality-wise it stayed the same. The new format is indeed shorter and simpler. Will update to new format.

File: vpn-daemon/main.go [hunk 5]
```diff
@@ -151,6 +152,49 @@ func handleClientConnection(clientConnection net.Conn) {
...
+			}
+
+			if err := os.MkdirAll(cnProfileDir, 0711); err != nil {
+				writer.WriteString(fmt.Sprintf("ERR: %s\n", err.Error()))
+				writer.Flush()
+				fmt.Println(err.Error())
+				continue
+			}
+
+			for _, profileName := range profileList {
```
Comment (fkooman) [inline vpn-daemon/main.go]: Nothing is ever written to a file?

File: vpn-daemon/main.go [hunk 6]
```diff
@@ -151,6 +152,49 @@ func handleClientConnection(clientConnection net.Conn) {
...
+				file, err := os.OpenFile(filepath.Join(cnProfileDir, profileName), os.O_CREATE, 0644)
+				if err != nil {
+					fmt.Println(err.Error())
+					continue
+					//XXX return error or just continue?
+					//for now we will skip the current profile
+				}
+
+				defer file.Close()
```
Comment (fkooman) [inline vpn-daemon/main.go]: `defer` cannot work here, as there is never a return...

**Comment by fkooman** [comment]
https://github.com/letsconnectvpn/vpn-daemon/compare/setup

What do you think of this branch for the `SETUP` command?

File: vpn-daemon/main.go (hunk 5 above)
Comment (auamod) [reply vpn-daemon/main.go]: For now, no. I opted to use just file-names, to avoid writing to files altogether.

I hope that it will remove the need to write/read the files and the possible issues related with it.

TLDR: 2 less possible errors

Comment (fkooman) [reply vpn-daemon/main.go]: At the cost of deleting folders and files? :-)
